libs/s25main/desktops/dskOptions.cpp

```diff
@@ -663,3 +688,91 @@ void dskOptions::loadVideoModes()
     // Sort by aspect ratio
     std::sort(video_modes.begin(), video_modes.end(), cmpVideoModes);
 }
+
+void dskOptions::Msg_ScreenResize(const ScreenResizeEvent& sr)
+{
+    Desktop::Msg_ScreenResize(sr);
+    updateGuiScale();
+}
+
+bool dskOptions::Msg_WheelUp(const MouseCoords& mc)
+{
+    if(!scrollGuiScale(true))
+        return Desktop::Msg_WheelUp(mc);
+    return true;
+}
+
+bool dskOptions::Msg_WheelDown(const MouseCoords& mc)
+{
+    if(!scrollGuiScale(false))
+        return Desktop::Msg_WheelDown(mc);
+    return true;
+}
+
+void dskOptions::updateGuiScale(ctrlComboBox* combo)
+{
+    // generate GUI scale percentages in 10% increments
+    constexpr auto stepSize = 10u;
+    const auto roundGuiScale = [=](unsigned percent) { return (percent + stepSize / 2) / stepSize * stepSize; };
+
+    unsigned minPercent, maxPercent, recommendedPercent;
+    VIDEODRIVER.getGuiScaleRange(&minPercent, &maxPercent, &recommendedPercent);
+    const auto recommendedPercentRounded = roundGuiScale(recommendedPercent);
+
+    if(!combo)
+    {
+        auto* group = GetCtrl<ctrlGroup>(ID_grpGraphics);
+        combo = group->GetCtrl<ctrlComboBox>(ID_cbGuiScale);
+    }
+
+    guiScales_.clear();
+    combo->DeleteAllItems();
+
+    combo->AddString((boost::format(_("Auto (%u%%)")) % recommendedPercent).str());
+    if(SETTINGS.video.guiScale == 0)
+        combo->SetSelection(0);
+
+    for(unsigned percent = roundGuiScale(minPercent); percent <= maxPercent; percent += stepSize)
+    {
+        if(percent == recommendedPercentRounded)
+            recommendedGuiScaleIndex_ = guiScales_.size();
+        guiScales_.push_back(percent);
+
+        combo->AddString(helpers::toString(percent) + "%");
+        if(percent == SETTINGS.video.guiScale)
+            combo->SetSelection(combo->GetNumItems() - 1);
+    }
+
+    // if GUI scale exceeds maximum, lower it to keep UI elements on screen
+    if(SETTINGS.video.guiScale > guiScales_.back())
+    {
+        combo->SetSelection(combo->GetNumItems() - 1);
+        SETTINGS.video.guiScale = guiScales_.back();
+        VIDEODRIVER.setGuiScale(SETTINGS.video.guiScale);
+    }
+}
+
+bool dskOptions::scrollGuiScale(bool up)
+{
+    if(!VIDEODRIVER.GetModKeyState().ctrl)
+        return false;
+
+    unsigned index = 0;
+    auto* group = GetCtrl<ctrlGroup>(ID_grpGraphics);
+    auto* combo = group->GetCtrl<ctrlComboBox>(ID_cbGuiScale);
+    const auto& selection = combo->GetSelection();
+    if(!selection || *selection == 0) // No selection or "Auto" item selected
+        index = recommendedGuiScaleIndex_ + 1;
+    else if(!up && *selection > 1) // Scrolling down and not at the first item
+        index = *selection - 1;
+    else if(up && *selection < guiScales_.size()) // Scrolling up and not at the last item
+        index = *selection + 1;                   // (no "- 1" needed because of "Auto" item)
+    else
+        return true;
```

Comment:
It might be clearer if instead of returning you wrap the change below into something like `if(index != selection)`, especially with the above suggested fix.

Reply:
(See above.)